File: hw/act1.3/python_approaches/backtracking.py
board = [[1, 0, 0, 0], [1, 1, 0, 1], [0, 1, 0, 0], [1, 1, 1, 1]]

# Backtracking Method 

def get_all_combinations(board: list[list[int]], start: list[int]) -> list[list[int]]:
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)] # all possible combinations 
    possible_dir = [] 
    for direction in directions:
        x_dir = start[0] + direction[0]
        y_dir = start[1] + direction[1] 

        if 0 <= x_dir < len(board) and 0 <= y_dir < len(board):
            path = board[x_dir][y_dir]
            if path == 1:
                possible_dir.append([x_dir, y_dir])
                        
    return possible_dir
        
def get_distance(start, dest) -> int:
    return sum([dest[0] - start[0], dest[1] - start[1]])

def get_best_choice(board, start, dest) -> list[list[int]]: 
    combinations = []
    combinations.append(start)
    while start != dest:
        
        combs = get_all_combinations(board, start)
        if not combs: 
            print("It is no possible to get to this destination")
            return 

        min_val = len(board) ** 10 # can't have -1 in how close it is 
        min_comb = None
        for comb in combs:
            if get_distance(comb, dest) < min_val:
                min_val = get_distance(comb, dest)
                min_comb = comb 
        start = min_comb
        combinations.append(start)
    return combinations

def print_board(board, combinations) -> None:

    for i in range(0, len(board)):
        for j in range(0, len(board)):
            if [i, j] in combinations:
                print(1, end = " ")
            else: 
                print(0, end = " ")
        print("\n")


def main(board):
    dest = [len(board) - 1, len(board) - 1] # [3, 3]
    start = [0, 0] # first elem 
    choices = get_best_choice(board, start, dest)
    # print(get_all_combinations(board, start))
    print_board(board, choices)
main(board)

File: hw/act1.3/backtracking.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

typedef pair<int, int> Point;

vector<Point> get_all_combinations(const vector<vector<int>>& board, const Point& start) {
    vector<Point> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    vector<Point> possible_dir;
    for (const auto& direction : directions) {
        int x_dir = start.first + direction.first;
        int y_dir = start.second + direction.second;
        if (x_dir >= 0 && x_dir < board.size() && y_dir >= 0 && y_dir < board[0].size()) {
            if (board[x_dir][y_dir] == 1) {
                possible_dir.push_back({x_dir, y_dir});
            }
        }
    }
    return possible_dir;
}

int get_distance(const Point& start, const Point& dest) {
    return abs(dest.first - start.first) + abs(dest.second - start.second);
}

vector<Point> backtrack(const vector<vector<int>>& board, const Point& start, const Point& dest, vector<vector<bool>>& visited) {
    if (start == dest) {
        return {start};
    }

    visited[start.first][start.second] = true;
    vector<Point> possible_moves = get_all_combinations(board, start);
    

    sort(possible_moves.begin(), possible_moves.end(), 
         [&dest](const Point& a, const Point& b) {
             return get_distance(a, dest) < get_distance(b, dest);
         });

    for (const auto& move : possible_moves) {
        if (!visited[move.first][move.second]) {
            vector<Point> path = backtrack(board, move, dest, visited);
            if (!path.empty()) {
                path.insert(path.begin(), start);
                return path;
            }
        }
    }

    visited[start.first][start.second] = false;
    return {};
}

vector<Point> get_best_choice(const vector<vector<int>>& board, const Point& start, const Point& dest) {
    vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size(), false));
    vector<Point> path = backtrack(board, start, dest, visited);
    
    if (path.empty()) {
        cout << "It is not possible to get to this destination" << endl;
    }
    
    return path;
}

void print_board(const vector<vector<int>>& board, const vector<Point>& path) {
    for (int i = 0; i < board.size(); ++i) {
        for (int j = 0; j < board[i].size(); ++j) {
            if (find(path.begin(), path.end(), Point{i, j}) != path.end()) {
                cout << "1 ";
            } else {
                cout << board[i][j] << " ";
            }
        }
        cout << endl;
    }
}

int main() {
    vector<vector<int>> board = {
        {1, 0, 0, 0},
        {1, 1, 0, 1},
        {0, 1, 0, 0},
        {1, 1, 1, 1}
    };

    Point dest = {static_cast<int>(board.size()) - 1, static_cast<int>(board[0].size()) - 1};
    Point start = {0, 0};

    vector<Point> choices = get_best_choice(board, start, dest);

    if (!choices.empty()) {
        print_board(board, choices);
    } else {
        cout << "It was not possible :(((" << endl;
    }

    return 0;
}